presto-hive/src/main/java/com/facebook/presto/hive/parquet/reader/ParquetReader.java

```diff
@@ -151,102 +155,193 @@ private boolean advanceToNextRowGroup()
+                        // Struct is defined and not empty
+                        structIsNull.add(false);
+                        offset++;
+                        structOffsets.add(offset);
+                    }
+                }
+            }
+            structDefinitionLevels.addAll(fieldDefinitionLevels);
+            structRepetitionLevels.addAll(fieldRepetitionLevels);
+        }
+    }
+
+    /**
+     * Each collection (Array or Map) has four variants of presence:
+     * 1) Collection is not defined, because one of it's optional parent fields is null
+     * 2) Collection is null
+     * 3) Collection is defined but empty
+     * 4) Collection is defined and not empty. In this case offset value is increased by the number of elements in that collection
+     */
+    private void calculateCollectionOffsets(String[] path, IntList offsets, BooleanList collectionIsNull, IntList definitionLevels, IntList repetitionLevels)
+    {
+        int maxDefinitionLevel = getMaxDefinitionLevel(path);
+        int maxElementRepetitionLevel = getMaxRepetitionLevel(path) + 1;
+        int offset = 0;
+        offsets.add(offset);
+        for (int i = 0; i < definitionLevels.size(); i = getNextCollectionStartIndex(repetitionLevels, maxElementRepetitionLevel, i)) {
+            if (isValueNull(isRequired(path), definitionLevels.get(i), maxDefinitionLevel)) {
+                //Collection is null
+                collectionIsNull.add(true);
+                offsets.add(offset);
+            }
+            else if (definitionLevels.get(i) == maxDefinitionLevel) {
+                // Collection is defined but empty
+                collectionIsNull.add(false);
+                offsets.add(offset);
+            }
+            else if (definitionLevels.get(i) > maxDefinitionLevel) {
+                //Collection is defined and not empty
+                collectionIsNull.add(false);
+                offset += getCollectionSize(repetitionLevels, maxElementRepetitionLevel, i + 1);
+                offsets.add(offset);
+            }
         }
+    }
 
-        int blockSize = blocks[0].getPositionCount();
-        int[] offsets = new int[blockSize + 1];
-        for (int i = 1; i < offsets.length; i++) {
-            elementOffsets.add(parameters.size());
-            offsets[i] = i;
+    private int getNextCollectionStartIndex(IntList repetitionLevels, int maxRepetitionLevel, int elementIndex)
+    {
+        do {
+            elementIndex++;
         }
-        return new RowBlock(0, blockSize, new boolean[blockSize], offsets, blocks);
+        while (hasMoreElements(repetitionLevels, elementIndex) && !isCollectionBeginningMarker(repetitionLevels, maxRepetitionLevel, elementIndex));
+        return elementIndex;
+    }
+
+    private int getCollectionSize(IntList repetitionLevels, int maxRepetitionLevel, int nextIndex)
+    {
+        int size = 1; // taking into account, that the method is called for not empty collections only
```

Comment:
@nezihyigitbasi it will be a bit complex assertion, because it requires to pass three additional parameters to getCollectionSize method (defenitionLevel, maxDefenitionLevel and isRequired  ). That's why I just added javadoc comment.